### PR TITLE
os/bluestore/NVMEDevice.cc: use slist for tracking mempool

### PR DESCRIPTION
**Pull request by optimistyzy**
Use the linked list is enough.

Signed-off-by: Ziye Yang <ziye.yang@intel.com>

## Checklist
- [ ] References tracker ticket
- [ ] Updates documentation if necessary
- [ ] Includes tests for new functionality or reproducer for bug

---

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard backend`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`

</details>

### CODE REVIEW

**Comment by tchaikov** [comment]
@optimistyzy just wanted to update you, i will try to catch up with your pull requests. have been struggling with other issues popped recently.

**Comment by optimistyzy** [comment]
> @optimistyzy just wanted to update you, i will try to catch up with your pull requests. have been struggling with other issues popped up recently.

Hi Kefu,

Thanks for your info. Those patches submitted by me are not urgent, please serve other urgent PRs first.

**Comment by optimistyzy** [comment]
@tchaikov If you have time, could you review this PR. In my platform, it proved the performance benefit.  Thanks.

File: src/os/bluestore/NVMEDevice.cc [hunk 1]
```diff
@@ -246,12 +254,16 @@ struct Task {
   }
   void release_segs(SharedDriverQueueData *queue_data) {
     if (io_request.extra_segs) {
-      for (uint16_t i = 0; i < io_request.nseg; i++)
-        queue_data->data_buf_mempool.push_back(io_request.extra_segs[i]);
+      for (uint16_t i = 0; i < io_request.nseg; i++) {
+        STAILQ_INSERT_HEAD(&queue_data->data_buf_list, (struct data_cache_buf *)io_request.extra_segs[i], link);
+        queue_data->data_buf_count++;
```
Comment (tchaikov) [inline src/os/bluestore/NVMEDevice.cc:228]:
```c++
auto buf = reinterpret_cast<data_cache_buf *>(io_request.extra_segs[i]);
queue_data->data_buf_list.push_front(*buf);
```

File: src/os/bluestore/NVMEDevice.cc [hunk 2]
```diff
@@ -334,7 +346,7 @@ int SharedDriverQueueData::alloc_buf_from_pool(Task *t, bool write)
   if (t->len % data_buffer_size)
     ++count;
   void **segs;
-  if (count > data_buf_mempool.size())
+  if (count > data_buf_count)
```
Comment (tchaikov) [inline src/os/bluestore/NVMEDevice.cc:318]:
```c++
if (count > data_buf_list.size())
```

File: src/os/bluestore/NVMEDevice.cc [hunk 3]
```diff
@@ -343,8 +355,11 @@ int SharedDriverQueueData::alloc_buf_from_pool(Task *t, bool write)
     segs = t->io_request.extra_segs;
   }
   for (uint16_t i = 0; i < count; i++) {
-    segs[i] = data_buf_mempool.back();
-    data_buf_mempool.pop_back();
+    ceph_assert(!(STAILQ_EMPTY(&data_buf_list)));
+    data_buf_count--;
+    segs[i] = STAILQ_FIRST(&data_buf_list);
+    STAILQ_REMOVE_HEAD(&data_buf_list, link);
+    ceph_assert(segs[i] != NULL);
```
Comment (tchaikov) [inline src/os/bluestore/NVMEDevice.cc:331]:
```c++
ceph_assert(!data_buf_list.empty());
segs[i] = &data_buf_list.front();
ceph_assert(segs[i] != nullptr);
data_buf_list.pop_front();
```

**Comment by optimistyzy** [comment]
> i'd suggest use slist from boost instead of using the macros from BSD. for better readability, and type safety.

Hi  Kefu,

Let me update the patch according to your comments

File: src/os/bluestore/NVMEDevice.cc [hunk 4]
```diff
@@ -78,6 +78,10 @@ struct IORequest {
   void **extra_segs = nullptr;
 };
 
+struct data_cache_buf {
+  STAILQ_ENTRY(data_cache_buf) link;
+};
```
[the remark this answers is not in the document]
Comment (optimistyzy) [reply src/os/bluestore/NVMEDevice.cc]: done

File: src/os/bluestore/NVMEDevice.cc [hunk 5]
```diff
@@ -143,7 +147,8 @@ class SharedDriverQueueData {
   public:
     uint32_t current_queue_depth = 0;
     std::atomic_ulong completed_op_seq, queue_op_seq;
-    std::vector<void*> data_buf_mempool;
+    STAILQ_HEAD(, data_cache_buf) data_buf_list;
```
[the remark this answers is not in the document]
Comment (optimistyzy) [reply src/os/bluestore/NVMEDevice.cc]: done

File: src/os/bluestore/NVMEDevice.cc [hunk 6]
```diff
@@ -162,14 +167,16 @@ class SharedDriverQueueData {
...
+    STAILQ_INIT(&data_buf_list);
     // allocate spdk dma memory
     for (uint16_t i = 0; i < data_buffer_default_num; i++) {
       void *b = spdk_dma_zmalloc(data_buffer_size, CEPH_PAGE_SIZE, NULL);
       if (!b) {
         derr << __func__ << " failed to create memory pool for nvme data buffer" << dendl;
         ceph_assert(b);
       }
-      data_buf_mempool.push_back(b);
+      STAILQ_INSERT_HEAD(&data_buf_list, (struct data_cache_buf *)b, link);
```
[the remark this answers is not in the document]
Comment (optimistyzy) [reply src/os/bluestore/NVMEDevice.cc]: done

File: src/os/bluestore/NVMEDevice.cc [hunk 7]
```diff
@@ -143,7 +147,8 @@ class SharedDriverQueueData {
   public:
     uint32_t current_queue_depth = 0;
     std::atomic_ulong completed_op_seq, queue_op_seq;
-    std::vector<void*> data_buf_mempool;
+    STAILQ_HEAD(, data_cache_buf) data_buf_list;
+    uint32_t data_buf_count = 0;
```
[the remark this answers is not in the document]
Comment (optimistyzy) [reply src/os/bluestore/NVMEDevice.cc]: done

File: src/os/bluestore/NVMEDevice.cc [hunk 8]
```diff
@@ -162,14 +167,16 @@ class SharedDriverQueueData {
     qpair = spdk_nvme_ctrlr_alloc_io_qpair(ctrlr, &opts, sizeof(opts));
     ceph_assert(qpair != NULL);
 
+    STAILQ_INIT(&data_buf_list);
```
[the remark this answers is not in the document]
Comment (optimistyzy) [reply src/os/bluestore/NVMEDevice.cc]: done

File: src/os/bluestore/NVMEDevice.cc [hunk 9]
```diff
@@ -162,14 +167,16 @@ class SharedDriverQueueData {
...
     // allocate spdk dma memory
     for (uint16_t i = 0; i < data_buffer_default_num; i++) {
       void *b = spdk_dma_zmalloc(data_buffer_size, CEPH_PAGE_SIZE, NULL);
       if (!b) {
         derr << __func__ << " failed to create memory pool for nvme data buffer" << dendl;
         ceph_assert(b);
       }
-      data_buf_mempool.push_back(b);
+      STAILQ_INSERT_HEAD(&data_buf_list, (struct data_cache_buf *)b, link);
+      data_buf_count++;
```
[the remark this answers is not in the document]
Comment (optimistyzy) [reply src/os/bluestore/NVMEDevice.cc]: done

File: src/os/bluestore/NVMEDevice.cc [hunk 10]
```diff
@@ -191,21 +198,22 @@ class SharedDriverQueueData {
...
-        void *b = data_buf_mempool[i];
-        ceph_assert(b);
-        spdk_dma_free(b);
-      }
-      data_buf_mempool.clear();
+    STAILQ_FOREACH_SAFE(b, &data_buf_list, link, tmp) {
+      STAILQ_REMOVE(&data_buf_list, b, data_cache_buf, link);
+      ceph_assert(b);
+      spdk_dma_free(b);
+      data_buf_count--;
```
[the remark this answers is not in the document]
Comment (optimistyzy) [reply src/os/bluestore/NVMEDevice.cc]: done

File: src/os/bluestore/NVMEDevice.cc (hunk 1 above)
Comment (optimistyzy) [reply src/os/bluestore/NVMEDevice.cc]: done

File: src/os/bluestore/NVMEDevice.cc (hunk 2 above)
Comment (optimistyzy) [reply src/os/bluestore/NVMEDevice.cc]: done

File: src/os/bluestore/NVMEDevice.cc (hunk 3 above)
Comment (optimistyzy) [reply src/os/bluestore/NVMEDevice.cc]: done

File: src/os/bluestore/NVMEDevice.cc [hunk 11]
```diff
@@ -38,6 +38,7 @@
 #include "common/errno.h"
 #include "common/debug.h"
 #include "common/perf_counters.h"
+#include <boost/intrusive/slist.hpp>
```
Comment (tchaikov) [inline src/os/bluestore/NVMEDevice.cc:41]: could you move this `#include` up before `#include <spdk/nvme.h>`. see https://google.github.io/styleguide/cppguide.html#Names_and_Order_of_Includes.

Comment (optimistyzy) [reply src/os/bluestore/NVMEDevice.cc:41]: Let update it.

**Pull request merged**